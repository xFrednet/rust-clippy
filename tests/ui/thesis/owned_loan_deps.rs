//@rustc-env: CLIPPY_PETS_PRINT=1

fn extend_string(data: &String) -> &String {
    data
}

#[forbid(clippy::borrow_pats)]
fn call_extend_simple() {
    let data = String::new();

    extend_string(&data).is_empty();
}

fn main() {}
