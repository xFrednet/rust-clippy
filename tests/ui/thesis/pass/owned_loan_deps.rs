//@rustc-env: CLIPPY_PETS_PRINT=1
//@rustc-env: CLIPPY_PRINT_MIR=1

fn extend_string(data: &String) -> &String {
    data
}

#[warn(clippy::borrow_pats)]
fn call_extend_simple() {
    let data = String::new();

    extend_string(&data).is_empty();
}

#[warn(clippy::borrow_pats)]
fn call_extend_named() {
    let data = String::new();
    let loan = &data;

    // The extention is not tracked for named loans
    extend_string(loan).is_empty();
}

fn debug() {
    let mut owned_a = String::from("=^.^=");
    let owned_b = String::from("=^.^=");
    let mut loan;

    loan = &owned_a;
    magic(loan);

    if true {
        owned_a.push('s');
    } else {
        magic(loan);
    }

    loan = &owned_b;
    magic(loan);
}

fn magic(_s: &String) {}

fn main() {}
