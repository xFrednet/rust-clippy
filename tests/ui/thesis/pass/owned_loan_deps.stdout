# "call_extend_simple"
- data           : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow, ArgBorrowExtended}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow}

# "call_extend_named"
- data           : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, NamedBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasNamedBorrow}

